rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala

```diff
@@ -1,23 +1,56 @@
 package coop.rchain.rholang.interpreter
 
-import cats.Monad
+import cats._
+import cats.data._
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
 import cats.implicits._
-import cats.mtl.MonadState
+import cats.mtl._
 
 import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
 
 package object accounting extends Costs {
 
-  type _cost[F[_]] = MonadState[F, Cost]
-  def _cost[F[_]](implicit ev: _cost[F]): _cost[F] = ev
+  type _cost[F[_]]    = MonadState[F, Cost]
+  type _costLog[F[_]] = FunctorTell[F, Chain[Cost]]
+
+  def _cost[F[_]](implicit ev: _cost[F]): _cost[F]          = ev
+  def _costLog[F[_]](implicit ev: _costLog[F]): _costLog[F] = ev
 
   def charge[F[_]: Monad](
       amount: Cost
-  )(implicit cost: _cost[F], error: _error[F]): F[Unit] =
+  )(implicit cost: _cost[F], error: _error[F], writer: _costLog[F]): F[Unit] =
     for {
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
+      _ <- writer.tell(Chain.one(amount))
       _ <- cost.modify(_ - amount)
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
     } yield ()
 
+  def costLog[M[_]: Sync](): FunctorListen[M, Chain[Cost]] =
+    new DefaultFunctorListen[M, Chain[Cost]] {
+      override val functor: Functor[M]  = implicitly[Functor[M]]
+      private val ref                   = Ref.unsafe(Chain.empty[Cost])
```

Comment:
We should avoid using Ref.unsafe. This makes costLog() and all its callers unsafe. 

At this level I would rather keep this abstract function safe by using `Ref.of` instead of `Ref.unsafe` and let the caller decide if he wants to play it safe or not.

Reply:
This implementation is currently used only in tests. I moved it there